src/pip/_internal/req/req_install.py

```diff
@@ -118,17 +117,22 @@ def __init__(
         self.req = req
         self.comes_from = comes_from
         self.constraint = constraint
-        if source_dir is None:
-            self.source_dir = None  # type: Optional[str]
-        else:
-            self.source_dir = os.path.normpath(os.path.abspath(source_dir))
         self.editable = editable
 
         if link is None and req and req.url:
             # PEP 508 URL requirement
             link = Link(req.url)
         self.link = self.original_link = link
         self.original_link_is_in_wheel_cache = False
+
+        self.source_dir = None  # type: Optional[str]
+        if self.editable:
+            assert link
+            if link.scheme == "file":
```

Comment:
For consistency with the conditional in the next block.

```suggestion
            if self.link.is_file:
```

Reply:
I'm OK with using `is_file`, that's a good change. But why use `self.link` rather than `link`? It's a property reference rather than a local variable, and locals are faster. Yes, I know premature optimisation, and it's not as if it's going to make any actual difference, but as a style thing, I generally prefer locals.

I'm not going to make a fuss either way, but just to note that I'm happy if @sbidoul wants to leave this as it was.

Reply:
For consistency with the next code block (i.e. the next conditional in this function) that does this.

https://github.com/pypa/pip/pull/8022/files#diff-47a4ece629e4698a5eaefea075d3cbdfR138

Reply:
Ok for `link.is_file`, done.

I prefer to keep `link` instead of `self.link` to keep this PR a pure refactoring without any behaviour change. Indeed `self.link` could be `req.url` if the `link` argument was `None`. So at least `assert link` must not be transformed to `assert self.link`.

Reply:
So I moved this code block before the assignment to `self.link`.

Reply:
Ah, indeed. Thanks for the clarification. :)